File: v1/2.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#     def myPrint(self):
#         print(self.val)
#         if self.next:
#             self.next.myPrint()


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        extra = 0
        head = ListNode(0)
        node = head
        while l1 or l2:
            data1 = l1.val if l1 else 0
            data2 = l2.val if l2 else 0
            total = data2 + data1 + extra
            if total > 9:
                extra = 1
                total = total - 10
            else:
                extra = 0
            node.next = ListNode(total)
            node = node.next
            l1, l2 = l1.next if l1 else None, l2.next if l2 else None
        if extra:
            node.next = ListNode(extra)
        return head.next



# if __name__ == "__main__":
#     list = ListNode(9)
#     list.next = ListNode(8)
#     print(Solution().addTwoNumbers(list, ListNode(1)).myPrint())

File: go/2.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

//type ListNode struct {
//	Val int
//	Next *ListNode
//	}

// 2 star
//方法很简单直接，但是有许多边角情况需要处理，容易漏掉或出错
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var extra, current int
	var left *ListNode
	p1, p2 := l1, l2
	dummy := &ListNode{Val: 0, Next:nil}
	p := dummy
	for {
		if p1 == nil || p2 == nil {
			break
		} else {
			current = p1.Val + p2.Val + extra
			if current > 9 {
				extra = 1
				current = current - 10
			} else {
				extra = 0
			}
			p1 = p1.Next
			p2 = p2.Next
		}
		p.Next = &ListNode{Val: current, Next:nil}
		p = p.Next
	}
	if p1 != nil || p2 != nil {
		if p1 != nil {
			left = p1
		} else {
			left = p2
		}
		for {
			if left == nil {
				break
			}
			current = left.Val + extra
			if current > 9 {
				current -= 10
				extra = 1
			} else {
				extra = 0
			}
			p.Next = &ListNode{Val: current, Next:nil}
			p = p.Next
			left = left.Next
		}

	}
	if extra > 0 {
		p.Next = &ListNode{Val: extra, Next:nil}
	}
	return dummy.Next

}
